katdal/h5datav2.py

```diff
@@ -457,6 +527,48 @@ def timestamps(self):
         extract_time = LazyTransform('extract_time', lambda t, keep: t + 0.5 * dump_period + time_offset)
         return LazyIndexer(self._timestamps, keep=self._time_keep, transforms=[extract_time])
 
+    def _vislike_indexer(self, dataset, extractor):
+        """Lazy indexer for vis-like datasets (vis / weights / flags).
+
+        This operates on datasets with shape (*T*, *F*, *B*) and potentially
+        different dtypes. The data type conversions are all left to the provided
+        extractor transform, while this method takes care of the common
+        selection issues, such as preserving singleton dimensions and dealing
+        with duplicate final dumps.
+
+        Parameters
+        ----------
+        dataset : :class:`h5py.Dataset` object or equivalent
+            Underlying vis-like dataset on which lazy indexing will be done
+        extractor : function, signature ``data = f(data, keep)``
+            Transform to apply to data (`keep` is user-provided 2nd-stage index)
+
+        Returns
+        -------
+        indexer : :class:`LazyIndexer` object
+            Lazy indexer with appropriate selectors and transforms included
+
+        """
+        # Create first-stage index from dataset selectors
+        time_keep = self._time_keep
+        # If there is a duplicate final dump, these lengths don't match -> ignore last dump in file
+        if len(time_keep) == len(dataset) - 1:
+            time_keep = np.zeros(len(dataset), dtype=np.bool)
+            time_keep[:len(self._time_keep)] = self._time_keep
+        stage1 = (time_keep, self._freq_keep, self._corrprod_keep)
+
+        def _force_3dim(data, keep):
```

Comment:
I can't see anything that this function is capturing from the outer scope, so it could be a class method rather than an inner function.